### PR TITLE
adds instanceProfileName field to nodegroups

### PR DESCRIPTION
**Pull request by jdavredbeard**
### Proposed changes
Add `instanceProfileName` as an `Input<string>` field to `NodeGroup` and `NodeGroupV2`. This would unblock https://github.com/pulumi/pulumi-self-hosted-installers/issues/181 by creating a path for users to create an `InstanceProfile` in a separate stack from the `NodeGroup` and pass the instance profile's name by stack reference. Currently https://github.com/pulumi/pulumi/issues/17515, prevents `NodeGroup` components from accepting `InstanceProfile` values created in another stack and provided to the `NodeGroup`'s stack by `aws.iam.InstanceProfile.get()`, and the `instanceProfile` parameter cannot accept an `InstanceProfile` resource passed by stack reference because the `instanceProfile` parameter is a plain `InstanceProfile` type, not an `Input<InstanceProfile>`.

### Related issues (optional)
Closes https://github.com/pulumi/pulumi-eks/issues/1497

### CODE REVIEW

File: nodejs/eks/cluster.ts [hunk 1]
```diff
@@ -802,7 +802,7 @@ export function createCore(
                 name,
                 parent,
                 args.instanceRole,
-                args.instanceProfileName,
```
Comment (flostadler) [inline nodejs/eks/cluster.ts:805]: We should do validation to ensure only one of them is set.

File: nodejs/eks/nodegroup.ts [hunk 2]
```diff
@@ -652,19 +653,54 @@ export function createNodeGroup(
     return createNodeGroupInternal(name, args, pulumi.output(core), parent, provider);
 }
 
+export function resolveOrGetInstanceProfile(
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:656]: Could you please add function docs explaining what this function is exactly doing and when it's throwing errors. Thanks!

File: nodejs/eks/nodegroup.ts [hunk 3]
```diff
@@ -652,19 +653,54 @@ export function createNodeGroup(
...
+        !c.nodeGroupOptions.instanceProfileName
+    ) {
+        throw new pulumi.ResourceError(
+            `an instanceProfile or instanceProfileName is required`,
+            parent,
+        );
+    }
+    if (
+        (args.instanceProfile || c.nodeGroupOptions.instanceProfile) &&
+        (args.instanceProfileName || c.nodeGroupOptions.instanceProfileName)
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:676]: You could simplify this with null coalescing:
```suggestion
        (args.instanceProfile?.instanceProfile && c.nodeGroupOptions?.instanceProfileName)
```

File: nodejs/eks/yarn.lock [hunk 4]
```diff
@@ -4,28 +4,28 @@
 
 "@ampproject/remapping@^2.2.0":
   version "2.3.0"
-  resolved "https://registry.yarnpkg.com/@ampproject/remapping/-/remapping-2.3.0.tgz#ed441b6fa600072520ce18b43d2c8cc8caecc7f4"
```
Comment (flostadler) [inline nodejs/eks/yarn.lock:7]: These seem to be unrelated changes

File: nodejs/eks/nodegroup.ts [hunk 5]
```diff
@@ -652,19 +653,54 @@ export function createNodeGroup(
...
+    ) {
+        throw new pulumi.ResourceError(
+            `invalid args for node group ${name}, instanceProfile and instanceProfileName are mutually exclusive`,
+            parent,
+        );
+    }
+    if (args.instanceProfile || c.nodeGroupOptions.instanceProfile) {
+        return args.instanceProfile ?? c.nodeGroupOptions.instanceProfile!;
+    }
+    return aws.iam.InstanceProfile.get(
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:686]: I think pulling in the resource with the `.get` function would incur RUM costs. We circumvent this by using the data source instead.

It seems like we're only interested in the ARN, right? Instance Profile ARNs follow this format: `arn:aws:iam::<account-id>:instance-profile/<profile-name>`.
If we know the name we can construct it without having to load the resource.

In that case we need other props, we could use [`iam.getInstanceProfileOutput`](https://www.pulumi.com/registry/packages/aws/api-docs/iam/getinstanceprofile/) instead.

File: provider/cmd/pulumi-gen-eks/main.go [hunk 6]
```diff
@@ -2308,7 +2308,11 @@ func nodeGroupProperties(cluster, v2 bool) map[string]schema.PropertySpec {
 				Ref:   awsRef("#/resources/aws:iam%2FinstanceProfile:InstanceProfile"),
 				Plain: true,
 			},
-			Description: "The ingress rule that gives node group access.",
+			Description: "The IAM InstanceProfile to use on the NodeGroup.",
+		},
+		"instanceProfileName": {
+			TypeSpec:    schema.TypeSpec{Type: "string"},
+			Description: "The name of the IAM InstanceProfile to use on the NodeGroup.",
```
Comment (flostadler) [inline provider/cmd/pulumi-gen-eks/main.go:2315]: We should point out here that `instanceProfile` and `instanceProfileName` is mutually exclusive.

**Review by flostadler — COMMENTED** [review]
Left some comments. Love the unit tests! But it would be great if we could add an integration test for this as well.

File: nodejs/eks/cluster.ts (hunk 1 above)
Comment (flostadler) [reply nodejs/eks/cluster.ts:805]: Ah saw you're doing that further down.

File: nodejs/eks/nodegroup.ts (hunk 3 above)
Comment (flostadler) [reply nodejs/eks/nodegroup.ts:676]: Mhm wait, my suggestion is not correct.

Aren't we susceptible to null pointer issues though? Is `c.nodeGroupOptions` guaranteed to be set?

Comment (flostadler) [reply nodejs/eks/nodegroup.ts:676]: Just checked, we're good. `c.nodeGroupOptions` is set even when we're not creating the default node group. All good!

File: nodejs/eks/nodegroup.ts [hunk 7]
```diff
@@ -653,42 +654,185 @@ export function createNodeGroup(
...
+ * Returns either the name of a passed InstanceProfile resource or a passed instanceProfileName.
+ * If an instanceProfile or instanceProfileName is passed in both the args and c.nodeGroupOptions,
+ * uses the input from args.
+ * Errors if no instanceProfile and no instanceProfileName is passed.
+ * Errors if both an instanceProfile and an instanceProfileName is passed.
+ */
+export function resolveInstanceProfileName(
     name: string,
     args: Omit<NodeGroupOptions, "cluster">,
-    c: pulumi.UnwrappedObject<CoreData>,
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts]: Why did you have to make this change?

File: nodejs/eks/nodegroup.ts [hunk 8]
```diff
@@ -698,10 +842,8 @@ function createNodeGroupInternal(
     parent: pulumi.ComponentResource,
     provider?: pulumi.ProviderResource,
 ): NodeGroupData {
-    const instanceProfile = core.apply((c) =>
-        resolveOrGetInstanceProfile(name, args, c, parent, provider),
-    );
-
+    const instanceProfileName = resolveInstanceProfileName(name, args, core, parent)
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:845]: Why not do `core.apply(c => resolveInstanceProfileName(...))`  instead like before?
That way resolveInstanceProfileName can stay out of the output-space and operate on promptly available data instead. That would simplify testing

File: nodejs/eks/nodegroup.ts [hunk 9]
```diff
@@ -653,42 +654,185 @@ export function createNodeGroup(
...
-        throw new pulumi.ResourceError(
-            `an instanceProfile or instanceProfileName is required`,
-            parent,
-        );
-    }
-    if (
-        (args.instanceProfile || c.nodeGroupOptions.instanceProfile) &&
-        (args.instanceProfileName || c.nodeGroupOptions.instanceProfileName)
-    ) {
+): pulumi.Output<string> | undefined {
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:811]: Shouldn't this always return a value?

File: nodejs/eks/nodegroup.ts (hunk 8 above)
Comment (flostadler) [reply nodejs/eks/nodegroup.ts:845]: The function would change to this in that case:

```ts
export function resolveInstanceProfileOrName(
    name: string,
    args: Omit<NodeGroupOptions, "cluster">,
    c: pulumi.UnwrappedObject<CoreData>,
    parent: pulumi.ComponentResource,
): pulumi.Output<string> {
    if (
        !args.instanceProfile &&
        !args.instanceProfileName &&
        !c.nodeGroupOptions.instanceProfile &&
        !c.nodeGroupOptions.instanceProfileName
    ) {
        throw new pulumi.ResourceError(
            `an instanceProfile or instanceProfileName is required`,
            parent,
        );
    }
    if (
        (args.instanceProfile || c.nodeGroupOptions.instanceProfile) &&
        (args.instanceProfileName || c.nodeGroupOptions.instanceProfileName)
    ) {
        throw new pulumi.ResourceError(
            `invalid args for node group ${name}, instanceProfile and instanceProfileName are mutually exclusive`,
            parent,
        );
    }

    return args.instanceProfile?.name ?? pulumi.output(c.nodeGroupOptions.instanceProfileName!) 
}
```

Comment (flostadler) [reply nodejs/eks/nodegroup.ts:845]: I'd also slightly change the definedness checks to check for all possible cases:
```ts
export function resolveInstanceProfileName(
    name: string,
    args: Omit<NodeGroupOptions, "cluster">,
    c: pulumi.UnwrappedObject<CoreData>,
    parent: pulumi.ComponentResource,
): pulumi.Output<string> {
    if (
        !args.instanceProfile || !c.nodeGroupOptions.instanceProfile ||
        !args.instanceProfileName || !c.nodeGroupOptions.instanceProfileName
    ) {
        // todo correct message that nothing's set
        throw new pulumi.ResourceError(
            `invalid args for node group ${name}, instanceProfile and instanceProfileName are mutually exclusive`,
            parent,
        );
    }

    if (args.instanceProfileName) {
        return pulumi.output(args.instanceProfileName);
    } else if (c.nodeGroupOptions.instanceProfileName) {
        return pulumi.output(c.nodeGroupOptions.instanceProfileName);
    } else if (args.instanceProfile !== undefined) {
        return args.instanceProfile.name;
    } else if (c.nodeGroupOptions.instanceProfile) {
        return c.nodeGroupOptions.instanceProfile.name;
    } else {
        throw new pulumi.ResourceError(
            `an instanceProfile or instanceProfileName is required`,
            parent,
        );
    }
}
```

File: tests/testdata/programs/self-managed-ng-instanceProfile/index.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+const ngv2_instanceProfileNameFromClusterNgOpts = new eks.NodeGroupV2("ng-instanceProfileNameFromClusterNgOpts", {
+  ...capacity,
+  cluster: cluster_instanceProfileNameFromNgOpts,
+  instanceType: "t4g.medium",
+  operatingSystem: eks.OperatingSystem.AL2023,
+});
+
+export const passedInstanceProfileName = instanceProfile.name
+export const kubeconfig = cluster_NoIp_NoIpn.kubeconfig;
```
[the remark this answers is not in the document]
Comment (jdavredbeard) [reply tests/testdata/programs/self-managed-ng-instanceProfile/index.ts:162]: Whoops missed exporting the other kubeconfigs... sure I can reduce the scope of the tests, I was going for a more completionist approach

**Review by flostadler — APPROVED** [review]
Nice! Thanks a lot
